controllers/services/auth/auth_controller_actions.go

```diff
@@ -0,0 +1,233 @@
+					newGroups = true
+				}
+			}
+		}
+
+		allowedg := foundGroups["allowedGroups"]
+		if len(allowedg) != 0 {
+			for _, e := range strings.Split(allowedg, ",") {
+				// If the group is already in the list then ignore it
+				if !slices.Contains(ai.Spec.AllowedGroups, e) {
+					ai.Spec.AllowedGroups = append(ai.Spec.AllowedGroups, strings.TrimSpace(e))
+					newGroups = true
+				}
+			}
+		}
+	}
+	if err != nil && !k8serr.IsNotFound(err) {
+		return err
+	}
+
+	// only update if we found a new group in the list
+	if newGroups {
+		err := rr.Client.Update(ctx, ai)
+		if err != nil {
+			return errors.New("error adding groups to Auth CR")
+		}
+	}
+
+	return nil
+}
+
+func bindRole(ctx context.Context, rr *odhtypes.ReconciliationRequest, groups []string, roleBindingName string, roleName string) error {
+	groupsToBind := []rbacv1.Subject{}
+	for _, e := range groups {
+		// we want to disallow adding system:authenticated to the adminGroups
+		if roleName == "admingroup-role" && e == "system:authenticated" {
+			log := logf.FromContext(ctx)
+			log.Info("system:authenticated cannot be added to adminGroups")
+			continue
+		}
+		rs := rbacv1.Subject{
+			Kind:     "Group",
+			APIGroup: "rbac.authorization.k8s.io",
+			Name:     e,
+		}
+		groupsToBind = append(groupsToBind, rs)
+	}
+
+	rb := &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleBindingName,
+			Namespace: rr.DSCI.Spec.ApplicationsNamespace,
+		},
+		Subjects: groupsToBind,
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "Role",
+			Name:     roleName,
+		},
+	}
+	err := rr.AddResources(rb)
+	if err != nil {
+		return errors.New("error creating RoleBinding for group")
+	}
+
+	return nil
+}
+
+func removeOldPermissions(ctx context.Context, rr *odhtypes.ReconciliationRequest, groups []string, roleBindingName string, roleName string) error {
+	roleBinding := &rbacv1.RoleBinding{}
+	if err := rr.Client.Get(ctx, client.ObjectKey{
+		Name:      roleBindingName,
+		Namespace: rr.DSCI.Spec.ApplicationsNamespace,
+	}, roleBinding); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	var groupsToRemove []rbacv1.Subject
+	for _, s := range roleBinding.Subjects {
+		var found = false
+		for _, g := range groups {
+			if s.Name == g {
+				found = true
+			}
+		}
+		if !found {
+			groupsToRemove = append(groupsToRemove, s)
+		}
+	}
+
+	if len(groupsToRemove) > 0 {
+		rb := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleBindingName,
+				Namespace: rr.DSCI.Spec.ApplicationsNamespace,
+			},
+			Subjects: groupsToRemove,
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: "rbac.authorization.k8s.io",
+				Kind:     "Role",
+				Name:     roleName,
+			},
+		}
+		err := rr.Client.Update(ctx, rb)
+		if err != nil {
+			return errors.New("error deleting RoleBinding for group")
+		}
+	}
+
+	return nil
+}
+
+func managePermissions(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
```

Comment:
I'm not sure I get how the `bindRole` and the `removeOldPermissions` are expected to be working, my understanding is:

1. a `RoleBinding` is created by invoking `bindRole` with some subjects populated i.e, from `ai.Spec.AdminGroups`
2. such `RoleBinding` is added to the resources the controller is going to deploy through the deploy action
3. the method `removeOldPermissions` is invoked, and the same set of groups i.e. `ai.Spec.AdminGroups` is passed to the method and if I got it right, it would remove the groups added in the step before.
4. the method `removeOldPermissions` directly updates the object, which is going to be overridden by the deploy action

I may be 100% incorrect :)

Reply:
No you are right. I was originally creating a rolebinding per group so that then needed to be cleaned up if the group was removed. I switched it to a single rolebinding encompassing all groups so that cleanup is no longer needed :facepalm: because the update handles it already. Removed